internal/testutils/logging.go

```diff
@@ -1,51 +1,132 @@
 package testutils
 
 import (
+	"bytes"
 	"fmt"
+	"io"
 	"os"
 	"os/exec"
+	"strings"
+	"sync"
+	"testing"
 	"time"
 )
 
+// TestWriter is an io.Writer that sends its output to a testing.T via t.Log,
+// ensuring logs are attributed to the correct test.
+// It buffers partial writes until a newline is seen.
+type TestWriter struct {
+	t   *testing.T
+	mu  sync.Mutex
+	buf bytes.Buffer
+}
+
+// NewTestWriter creates a new TestWriter that logs to t.
+func NewTestWriter(t *testing.T) *TestWriter {
+	return &TestWriter{t: t}
+}
+
+// Write implements io.Writer. It buffers until a newline is seen,
+// then flushes complete lines to t.Log. Remainders are held until
+// the next Write call.
+func (w *TestWriter) Write(p []byte) (n int, err error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	n, err = w.buf.Write(p)
+	for {
+		line, errLine := w.buf.ReadString('\n')
+		if errLine == io.EOF {
+			// leftover data without newline — keep it for next Write
+			w.buf.Reset()
+			w.buf.WriteString(line)
+			break
+		}
+		// Trim the newline, let t.Log add its own
+		w.t.Log(strings.TrimSuffix(line, "\n"))
+	}
+	return
+}
+
 // RunWithTiming runs the given command while logging its duration with the provided message.
-func RunWithTiming(msg string, cmd *exec.Cmd) error {
-	LogCommand(msg, cmd)
+//
+//nolint:thelper // we do call t.Helper() if t is not nil
+func RunWithTiming(t *testing.T, msg string, cmd *exec.Cmd) error {
+	if t != nil {
+		t.Helper()
+	}
+	w := testWriterOrStderr(t)
+
+	LogCommand(t, msg, cmd)
 	start := time.Now()
 	err := cmd.Run()
 	duration := time.Since(start)
 
 	if err != nil {
-		fmt.Fprintf(os.Stderr, redSeparatorf("%s failed in %.3fs with %v", msg, duration.Seconds(), err)+"\n")
+		fmt.Fprintf(w, redSeparatorf("%s failed in %.3fs with %v", msg, duration.Seconds(), err)+"\n")
 	} else {
-		fmt.Fprintf(os.Stderr, separatorf("%s finished in %.3fs", msg, duration.Seconds())+"\n")
+		fmt.Fprintf(w, separatorf("%s finished in %.3fs", msg, duration.Seconds())+"\n")
 	}
 	return err
 }
 
 // LogCommand logs the given command to stderr.
-func LogCommand(msg string, cmd *exec.Cmd) {
+//
+//nolint:thelper // we do call t.Helper() if t is not nil
+func LogCommand(t *testing.T, msg string, cmd *exec.Cmd) {
+	if t != nil {
+		t.Helper()
+	}
+	w := testWriterOrStderr(t)
+
 	sep := "----------------------------------------"
-	fmt.Fprintf(os.Stderr, "\n"+separator(msg)+"command: %s\n%s\nenvironment: %s\n%s\n", cmd.String(), sep, cmd.Env, sep)
+	fmt.Fprintf(w, "\n"+separator(msg)+"command: %s\n%s\nenvironment: %s\n%s\n", cmd.String(), sep, cmd.Env, sep)
```

Comment:
I thin we've an issue here, I'm getting this output:

```
===== Building authd =====
command: %s
%s
environment: %s
%s
```

Reply:
Indeed, the "Address linter issues" commit I pushed yesterday (after rebasing on main and getting new linter warnings from the new golangci-lint version) broke that output. I fixed that now.